Adding client documentation.

yupinghu opened the pull request.

yupinghu committed:
Add client documentation.

ericvw commented on README.md:
```diff
@@ -5,9 +5,11 @@ HTTP wrapper for [Bloomberg Open API].
 
 Find source code in the [GitHub repository].
 
+For information on how to use the HTTP wrapper, see the [Client Guide].
```
s/wrapper/api/ and should we call this "Client Guide" or API Protocol Guide (something along those lines)?

yupinghu replied:
Yeah, wasn't sure what to name this -- I was trying to use different terminology to distinguish between the HTTP API and the underlying Bloomberg Open API. API Protocol Guide is weird because it could be confused with the latter, but I am not happy with "Client Guide" either.

yupinghu replied:
Done.

ericvw commented on clientGuide.md:
```diff
@@ -0,0 +1,100 @@
+============
+
+The [Bloomberg Open API] provides access to market data. The
+[Bloomberg Open API HTTP Wrapper] makes the API available via HTTP: clients
```
Should we rename `Wrapper` with `Proxy` or `Server`?

yupinghu replied:
Renamed this a bit as discussed in person.

ericvw commented on clientGuide.md:
```diff
@@ -0,0 +1,100 @@
+
+The HTTP wrapper is currently in development, and does not yet provide access
+to all features of the API. Notably, it currently supports the
+Request/Response paradigm of the API, but not Subscription.
```
"supports the reference data service (i.e., /blp/refdata)"

I wouldn't mention what it doesn't support unless we are going to enumerate all the services note supported.

yupinghu replied:
Done.

ericvw commented on clientGuide.md:
```diff
@@ -0,0 +1,100 @@
+To request data, a client makes an HTTP POST request on a URL of the following
+form:
+
+http://blpapihost.example.com/v1.0/request/blp/refdata/{operation}
```
Does it make sense to introduce the general form of the URL first before the concrete examples?  I'm not torn on this, but just something to consider.

yupinghu replied:
As discussed, leaving as-is for now, but will definitely expand to more general forms when we support more stuff in future versions.

ericvw commented on clientGuide.md:
```diff
@@ -0,0 +1,100 @@
+message = "OK" but the data returned from the API will contain an error (in
+a responseError property of the data, or a securityError property of one or
+more of the securityData).
+
```
It may be a good idea to have a simple response as an example here.

yupinghu replied:
Done.

ericvw commented:
CC'ing @ericlu88 for comments as well.

ericlu88 commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,196 @@
+Requesting data
+---------------
+
+To request data, a client makes an HTTP POST request on a URL of the following
```
HTTP => HTTPS?

ericvw replied:
"an HTTP" => "a HTTP"

ericlu88 commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,196 @@
+
+Let's examine each segment of the path:
+
+* /v1.0 -- The version of the HTTP API we're using.
```
As part of the refactoring yesterday, I remove the **/v1.0** out of the url. Sorry for the late reminder.

ericlu88 commented:
Some other enhancement that I added into the api:
- The server now understand **gzip** encoding. The client may specify the behavior by adding `accept-encoding: gzip`. For curl client, they can use `--compressed` to enable this feature.
- We removed the versioning from the URL but baked into the `accept-version` header. The server is running **@1.0.0** now. We strongly encourage clients to specify the version they want to hit, but the default behavior is to default to version **1.0** if the header is omitted. Couple of example of specifying the version:
  - `accept-version: 1`
  - `accept-version: 1.0`
  - `accept-version: ~1`
- The `accept` header specifies the protocol that the client wants the server to respond with. Currently the server only understands how to respond in **json**, so the accept header has to contain that at least. By default curl is sending `accept: */*` so the user don't have to worry about it.

yupinghu commented:
- Switched to https in the docs.
- Removed v1.0 in URLs.
- Added a section for the request headers.

Please take a look. I'm going to add more to the bottom sections (i.e. probably starting with apiflds now that I can generate example output) in the meantime.

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,206 @@
+
+This document provides an overview of the [HTTP API]: the URLs to use and the
+formatting for the HTTP requests and responses. We assume, for the purposes of
+this doc, that the server at http://blpapihost.example.com/ is running an
```
s/implementation/instance

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,206 @@
+this doc, that the server at http://blpapihost.example.com/ is running an
+implementation of this API.
+
+This doc also includes basic info about some of the underlying [Open API]
```
s/doc/document && s/info/information

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,206 @@
+
+* /request -- The Open API paradigm we're using, namely Request/Response.
+* /blp/refdata -- The service on which we're making the request. Currently /blp/refdata and /blp/apiflds are supported; more will be added in future versions.
+* &lt;operation&gt; -- The operation for this request. For /blp/refdata, the operations are listed in the Developer's Guide in section A.2.1 (and detailed in 7.2). Note that the URL omits the "Request" in the operation name, e.g. for the HistoricalDataRequest operation, use /HistoricalData
```
Should `Developer's Guide` be markdown linked?

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,206 @@
+--------------------
+
+The client may specify the following HTTP request headers to tell the server
+what sorts of responses it can use:
```
It may be a good idea to partition which headers are required and optional. (i.e., Content-Type:) is more or less required.  @ericlu88 can you confirm?

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,206 @@
+* message: If we successfully communicated with the Open API, this will be "OK". Otherwise, it will contain a description of the error.
+* data: The response from the Open API (only present if message = "OK"). The exact structure of this value depends on the operation, and is detailed in section A.2.14 of the [Developer's Guide].
+
+Note that message = "OK" doesn't necessarily mean that the request successfully
```
Change to: `"message": "OK"`

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,206 @@
+Below is the response from the example request in the previous section. Notice
+that the response JSON at the top level has just the three properties, and that
+the data property is where all the interesting stuff is happening. For more
+details about the shape of HistoricalData responses, consult section A.2.5 in
```
s/shape/structure

ericvw commented on HTTP_API_Guide.md:
````diff
@@ -0,0 +1,206 @@
+```
+
+
+Reference Data Service (blp/refdata)
````
//blp/refdata

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,206 @@
+
+The Reference Data Service provides request/response access to market data. It
+supports several kinds of requests, each with its own request parameters and
+response types. The sections below describe the more interesting/useful types
```
The sub-sections below describe the different types of requests/responses that are available via the [HTTP API].

ericvw commented on HTTP_API_Guide.md:
````diff
@@ -0,0 +1,206 @@
+Example request/response:
+
+```
+curl $OPTS -X POST https://blpapihost.example.com/request/blp/refdata/ReferenceData?sessid=$sessid --data @- <<EOF
````
Remove `$OPTS`

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,206 @@
+
+There are more fields available than can be enumerated in this document.
+To search for and get information about fields using a Bloomberg Terminal,
+use FLDS<GO>. You can also use the HTTP API to query the /blp/apiflds service.
```
Inline code block (i.e., `) FLDS<GO>

yupinghu commented:
I believe I've addressed all the feedback so far. I'm now trying to figure out how FieldSearch works (specifically the exclude/include options), so that'll be the next section, although at this point, I think we could probably go with what we have.

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,256 @@
+
+The [Bloomberg Open API] provides access to market data. The
+[Bloomberg HTTP API] makes the [Open API] available via HTTP: clients POST
+commands to a HTTP server, which will respond with the relevant data.
```
s/relevant/corresponding

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,256 @@
+[Bloomberg HTTP API] makes the [Open API] available via HTTP: clients POST
+commands to a HTTP server, which will respond with the relevant data.
+
+The [HTTP API] is currently in development, and does not yet provide access
```
Remove `,`.  `development and does ...`

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,256 @@
+commands to a HTTP server, which will respond with the relevant data.
+
+The [HTTP API] is currently in development, and does not yet provide access
+to all features of the [Open API]. At this time it supports the Reference Data
```
"to all services that the [Open API] provides."

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,256 @@
+
+This document also includes basic information about some of the underlying
+[Open API] operations. For more details on the [Open API], refer to the
+[BLPAPI Developer's Guide]. This document will often refer to specific sections
```
"Where appropriate, this document will..."

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,256 @@
+Requesting data
+---------------
+
+To request data, a client makes a HTTP POST request on a URL of the following
```
s/on/to

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,256 @@
+`/blp/refdata` and `/blp/apiflds` are supported; more will be added in future
+versions.
+* `<operation>` -- The operation for this request. For `/blp/refdata`,
+the operations are listed in the [Developer's Guide] in section A.2.1 (and
```
Switch sections and reference to the guide.  e.g., "the operations are listed in section A.2.1 of the [Developer's Guide]."

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,256 @@
+e.g. for the HistoricalDataRequest operation, use `/HistoricalData`
+
+The body of the POST should consist of a JSON object containing the parameters
+of the request. Each operation defines what parameters it uses; sections 7.2
```
s/of/for

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,256 @@
+and A.2 of the [Developer's Guide] provide full details.
+
+As an example, the following will request the open and last price for IBM and
+Apple for each day during the period Jan 1 - 5, 2012.
```
Apple "stock" for each day

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,256 @@
+of one or more of the `securityData`).
+
+Below is the response from the example request in the previous section. Notice
+that the response JSON at the top level has just the three properties, and that
```
Remove `,`

ericvw commented on HTTP_API_Guide.md:
```diff
@@ -0,0 +1,256 @@
+
+The Reference Data Service provides request/response access to market data. It
+supports several kinds of requests, each with its own request parameters and
+response types. The sub-sections below describe the different tyles of
```
s/tyles/types

ericvw commented:
This lgtm.  @yupinghu can you squash these down?

yupinghu commented:
Squashed down.
